# Data Understanding

## Objective

The objective of this step is to explore the dataset structure, understand the features, and identify potential data quality issues before performing data preprocessing and modeling.

In [1]:
import pandas as pd

In [2]:
file_name = '../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv'

In [3]:
df = pd.read_csv(file_name)

## Dataset Overview

In this section, we explore the general structure of the dataset, including the number of rows, columns and available features.

In [4]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [5]:
df.shape

(7043, 21)

### Dataset Size

The dataset contains 7043 customer records and 21 columns.

Each row represents a customer and each column contains information about customer characteristics, services, account details or churn status.

In [6]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

### Feature Description

The dataset contains customer demographic information, subscribed services, account details and the target variable Churn.

The features include numerical, categorical and binary variables that describe customer characteristics and service usage.

In [8]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


### Numerical Features Analysis

The numerical features were examined using descriptive statistics.

- `SeniorCitizen` is a binary feature and does not contain meaningful numerical outliers.
- `tenure` values are within a reasonable range (0 to 72 months).
- `MonthlyCharges` values are within a reasonable range and no significant outliers were detected based on statistical boundaries.

In [9]:
df.describe(include='str')

,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,7590-VHVEG,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,20.2,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


## Initial Data Quality Observations

During the initial exploration of the dataset, some potential data quality issues were idetified:

- `customerID` is an identifier column and does not provide meaningful predictive information. It may be removed during feature engineering.
- `TotalCharges` represents a numerical feature, but it is stored as a string data type and requires conversion before modeling.

In [10]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

## Missing Values Check

No missing values were detected using standard missing value detection methods.

However, the `TotalCharges` feature requires further investigation because it is stored as a string data type and may contain non-numeric values.

In [11]:
df['Churn'].value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

## Target Variable Distribution

The target variable `Churn` is moderately imbalanced.

Approximately 73.4% of customers have not churned, while 26.5% of customers have churned.

Because the dataset is imbalanced, accuracy alone may not be a reliable evaluation metric. Therefore additional metrics such as precision, recall and F1-score will be considered.

# Summary of Data Understanding

## Key Findings

- The dataset contains 7043 customer records and 21 columns.
- The target variable is `Churn`, which indicates whether a customer has left the service.
- The dataset contains numerical, categorical, and binary features.
- `customerID` is an identifier column and is not expected to provide predictive value.
- `TotalCharges` is a numerical feature stored as a string and requires data type conversion during preprocessing.
- No standard missing values were detected.
- The target variable is moderately imbalanced, with approximately 73% non-churned customers and 26% churned customers.
- Numerical features were reviewed, and no significant outliers were identified.

## Next Steps

The next phase of the project is Data Cleaning, where the raw dataset will be cleaned and ready for further analysis and modeling.